File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        maxArea = 0
        height = len(grid)
        width = len(grid[0])
        
        moving = [[0,1],[0,-1],[1,0],[-1,0]]
        
        def solve_dfs(x, y) -> int:
            
            if grid[x][y] == 0: return 0
            
            grid[x][y] = 0
            
            retArea = 1
            
            for mov in moving:
                nx = x + mov[0]
                ny = y + mov[1]
                
                if nx < 0 or ny < 0 or nx >= height or ny >= width:
                    continue
                
                retArea += solve_dfs(nx, ny)
            
            return retArea
            
        for i in range(height):
            for j in range(width):
                if grid[i][j] == 1:
                    maxArea = max(maxArea, solve_dfs(i, j))        

        return maxArea

File: 695-max-area-of-island/695-max-area-of-island.cpp
inline int MAX(int a,int b) {return a>b ? a:b;}
inline int MIN(int a,int b) {return a<b ? a:b;}
class Solution {
private:
    int dir[4][2] = { {1,0},{-1,0},{0,1},{0,-1} };
    int dfs(vector<vector<int>>& grid, int x, int y) {
        if (MIN(x,y) < 0 || x >= grid.size() || y >= grid[0].size() || grid[x][y] == 0) return 0;
        grid[x][y] = 0;
        int area = 1;
        for (int k = 0; k < 4; k++) {
            int nx = x + dir[k][0];
            int ny = y + dir[k][1];
            area += dfs(grid, nx, ny);
        }
        return area;
    }
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int answer = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j])
                    answer = MAX(answer,dfs(grid,i,j));
            }
        }
        return answer;
    }
};